### ISSUE TITLE
One strange game from SF-zh

### ISSUE THREAD

**Issue by Vinvin20** [ddugovic/Stockfish#97]
See here : http://www.talkchess.com/forum/viewtopic.php?p=691855#691855
" Stockfish announces mate in 21 at move 26th, next move the eval is +4.87 only and keeps going down."

**Comment by ddugovic** [comment]
For convenience and redundancy, here's the PGN:

```
[Event "My Tournament"] 
[Site "PhenomII"] 
[Date "2016.11.01"] 
[Round "1"] 
[White "Stockfish 011116"] 
[Black "Imortal 2.9001"] 
[Result "0-1"] 
[PlyCount "67"] 
[Termination "adjudication"] 
[TimeControl "900+3"] 
[Variant "crazyhouse"] 

1. d4 {book} e6 {-0.43/16 45s} 2. Nf3 {+0.15/22 47s} Nc6 {-0.39/15 43s} 
3. Nbd2 {+0.33/20 31s} Nf6 {-0.22/14 41s} 4. e4 {+0.59/18 12s} d6 {0.00/14 39s} 
5. Bb5 {+0.66/19 29s} Bd7 {-0.40/15 37s} 6. O-O {+0.80/18 8.6s} 
Be7 {-0.02/14 35s} 7. c3 {+1.20/19 16s} Rb8 {-0.47/13 34s} 8. Re1 {+1.25/20 39s} 
a6 {-0.21/14 27s} 9. Bd3 {+1.35/20 16s} Ng4 {-0.42/15 29s} 
10. h3 {+1.59/19 8.5s} Nxf2 {+0.04/14 19s} 11. Kxf2 {+2.90/16 8.8s} 
h5 {-0.62/12 28s} 12. Rh1 {+2.27/20 83s} Bh4+ {-0.25/12 22s} 
13. g3 {+3.09/16 8.0s} P@e3+ {-0.56/13 27s} 14. Kxe3 {+3.57/18 8.1s} 
Bxg3 {-1.57/14 26s} 15. Qg1 {+3.64/21 55s} Qf6 {-1.53/12 24s} 
16. Nf1 {+3.83/20 45s} Bh4 {-1.11/11 23s} 17. Nxh4 {+3.75/20 22s} 
Qxh4 {-2.85/10 21s} 18. N@f3 {+4.43/19 11s} Qe7 {-3.53/11 22s} 
19. Qxg7 {+5.64/17 12s} Qf8 {-5.04/10 21s} 20. Qxf8+ {+5.81/16 8.1s} 
Rxf8 {-2.73/11 20s} 21. P@g2 {+5.91/17 43s} e5 {-1.27/9 19s} 
22. Q@e2 {+6.00/16 24s} N@f4 {+0.78/9 12s} 23. dxe5 {+5.74/17 29s} 
Nxe2 {+0.19/9 18s} 24. Kxe2 {+5.90/17 32s} P@g4 {-0.02/9 13s} 
25. Ne1 {+6.88/16 23s} Nxe5 {-2.65/9 16s} 26. N@g7+ {+M21/17 49s} 
Kd8 {-0.57/10 14s} 27. B@f6+ {+4.87/19 45s} Kc8 {-1.49/11 15s} 
28. Bxe5 {+5.91/17 25s} dxe5 {+0.90/9 15s} 29. P@e7 {+7.23/17 38s} 
P@f3+ {+1.03/7 14s} 30. Kd1 {+6.14/16 35s} B@e2+ {+9.34/8 11s} 
31. Bxe2 {+1.01/16 22s} fxe2+ {+9.64/8 13s} 32. Kxe2 {-M26/16 15s} 
Bb5+ {+7.59/8 13s} 33. c4 {-M20/20 16s} Bxc4+ {+11.45/8 10s} 
34. N@d3 {-M12/20 12s, Black wins by adjudication} 0-1
```

**Comment by ddugovic** [comment]
Attempted to duplicate and encountered a segmentation fault!

```
setoption name UCI_Variant value crazyhouse
position fen 1r2kr2/1ppb1p2/p2p4/4n2p/4P1p1/2PB3P/PP2K1P1/R1B1NN1R/NPBPpqq w - - 50 26
go infinite
```

**Comment by ddugovic** [comment]
Could not duplicate mate evaluation with:

```
uci
setoption name UCI_Variant value crazyhouse
position fen 1r2kr2/1ppb1p2/p2p4/4n2p/4P1p1/2PB3P/PP2K1P1/R1B1NN1R[NPBPpqq] w - - 50 26
go infinite
```

**Comment by ianfab** [comment]
I am pretty sure that this is caused by hash collisions, since we have not included the pieces in hand into the calculation of the hash keys yet. This would also explain why you can not reproduce it.

**Comment by ddugovic** [comment]
Ah, that makes sense.

Now I'm shocked that Stockfish plays so well! :-)

**Comment by isaacl** [comment]
> since we have not included the pieces in hand into the calculation of the hash keys yet

In that case, is it too early for bug reports about player strength, etc?

**Comment by ddugovic** [comment]
I'm not sure, but in any case it would be helpful if duplication steps were provided in the issue description (in this case I had to go to a forum and figure out which PGN / FEN to use for testing).

**Comment by ianfab** [comment]
I have opened PR #99 that should fix hashing in crazyhouse.

**Comment by ddugovic** [comment]
Merged #99 and deploying to http://dugovic.mooo.com/

**Comment by niklasf** [comment]
Lichess should have these changes tomorrow (on client as well as server).
